### PR TITLE
Miniapp triangular solver

### PR DESCRIPTION
**Pull request by albestro**
Introduce miniapp for benchmarking the triangular solver.

The starting point is a
- left side
- lower
- non transpose
- non unit diagonal

and if not necessary, I would not expose these options.

It's just a starting point for now, I have to work on it.

### CODE REVIEW

File: miniapp/miniapp_triangular_solver.cpp [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  // options
+  using namespace hpx::program_options;
+  options_description desc_commandline("Usage: " HPX_APPLICATION_STRING " [options]");
+
+  // clang-format off
+  desc_commandline.add_options()
+    ("matrix-size",       value<SizeType>()->default_value(4096),  "Matrix size.")
+    ("block-size",        value<SizeType>()->default_value(256),   "Block cyclic distribution size.")
+    ("result-cols",       value<SizeType>()->default_value(512),   "Matrix size.")
+    ("result-block-cols", value<SizeType>()->default_value(512),   "Block cyclic distribution size.")
```
Comment (Sely85) [inline miniapp/miniapp_triangular_solver.cpp:148]: What about adding the matrix the values referred to? For example "Matrix A size" and "Matrix b size", or "Input matrix size" and "Output matrix size".

**Review by teonnik — COMMENTED** [review]
LGTM, except:

```
add_executable(miniapp_triangular_solver miniapp_triangular_solver.cpp)
target_link_libraries(miniapp_triangular_solver PRIVATE DLAF DLAF_test)
target_add_warnings(miniapp_triangular_solver)
```

the dependency on `DLAF_test` should be remove. The function `getLeftTriangularSystem()` can be moved to a header in `include/dlaf`.

A question I have is whether we would want to extend this miniapp to support other combinations of Side, Uplo and Transpose in the future? Currently, only LLN is distributed, once we implement the other variants, would we want separate miniapps for them or would we want to integrate them in this one? If it is the former, I suggest that we append the file name with  `_LLN`.

File: miniapp/miniapp_triangular_solver.cpp [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+  double m = A.size().rows();
+  double n = b.size().cols();
+  auto add_mul = n * m * m / 2;
+  const double total_ops = dlaf::total_ops<T>(add_mul, add_mul);
+
+  using dlaf::matrix::test::getLeftTriangularSystem;
+  std::function<T(const GlobalElementIndex&)> setter_A, setter_b, expected_b;
+  std::tie(setter_A, setter_b, expected_b) =
+      getLeftTriangularSystem<GlobalElementIndex, T>(uplo, op, diag, alpha, A.size().rows());
```
Comment (rasolca) [inline miniapp/miniapp_triangular_solver.cpp:88]: I suspect that setup will hit the subnormal problem as well.

File: miniapp/miniapp_triangular_solver.cpp [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+int hpx_main(hpx::program_options::variables_map& vm) {
+  options_t opts = check_options(vm);
+
+  Communicator world(MPI_COMM_WORLD);
+  CommunicatorGrid comm_grid(world, opts.grid_rows, opts.grid_cols, Ordering::ColumnMajor);
+
+  // Allocate memory for the matrices
+  MatrixType A(GlobalElementSize{opts.m, opts.m}, TileElementSize{opts.mb, opts.mb}, comm_grid);
```
Comment (rasolca) [inline miniapp/miniapp_triangular_solver.cpp:71]: naming conventions: `a`

**Review by rasolca — CHANGES_REQUESTED** [review]
Looks good as a first attempt, but we should pay attention to subnormal ops.

File: miniapp/miniapp_triangular_solver.cpp (hunk 1 above)
Comment (albestro) [reply miniapp/miniapp_triangular_solver.cpp]: I renamed the parameters according to the `trsm` function, so all the sizes refers to the B matrix, from which it is possible to infer all other sizes.

**Comment by albestro** [comment]
> the dependency on `DLAF_test` should be remove. The function `getLeftTriangularSystem()` can be moved to a header in `include/dlaf`.

As discussed separately, I reduced at maximum the dependency with `DLAF_test`:
- copied a slightly less general version of the function `getLeftTriangularSystem()` which generates the sample triangular system (now it works just with floating point values and not complex, so it does not depend anymore on `TypeUtilities` for generalized value type management)
- copied in-place the computation of mul/add error (from `TypeUtilties`)

Unfortunately, it was not easily possible to remove the last dependency because checking results requires `CHECK_MATRIX_NEAR` (that it is quite trivial and slack, it may be improved in the future). I added a comment as a reminder.

> A question I have is whether we would want to extend this miniapp to support other combinations of Side, Uplo and Transpose in the future? Currently, only LLN is distributed, once we implement the other variants, would we want separate miniapps for them or would we want to integrate them in this one? If it is the former, I suggest that we append the file name with `_LLN`.

If you agree, I would consider this in the future as soon as we will implement other variants.

File: miniapp/miniapp_triangular_solver.cpp (hunk 2 above)
Comment (albestro) [reply miniapp/miniapp_triangular_solver.cpp]: Experimentally, checking with `perf` for subnormal interrupts and by looking at single node performances, it does not suffer (with m=10240, n=10240). I'm checking how values of the reference linear systems are generated (I will share with you the details) and we should be on the safe side.

**Review by rasolca — APPROVED** [review]

**Pull request merged**